pkg/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,253 @@
+// stopped, the Event will not be enqueued, and its cancel channel will be
+// closed, indicating that the Event was not ran. This function may block if
+// there is an event being processed by the queue.
+func (q *EventQueue) Enqueue(ev *Event) <-chan interface{} {
+
+	select {
+	case <-q.close:
+		close(ev.cancelled)
+		close(ev.eventResults)
+		return nil
+	default:
+		// The events channel may be closed even if an event has been pushed
+		// onto the events channel, as events are consumed off of the events
+		// channel asynchronously! If the EventQueue is closed before this
+		// event is processed, then it will be cancelled.
+
+		ev.stats.waitEnqueue.Start()
+		q.events <- ev
+		ev.stats.waitEnqueue.End(true)
+		ev.stats.waitConsumeOffQueue.Start()
+		return ev.eventResults
+	}
+}
+
+func (ev *Event) printStats() {
+	log.WithFields(logrus.Fields{
+		"eventType":                    reflect.TypeOf(ev.Metadata).String(),
+		"eventHandlingDuration":        ev.stats.durationStat.Total(),
+		"eventEnqueueWaitTime":         ev.stats.waitEnqueue.Total(),
+		"eventConsumeOffQueueWaitTime": ev.stats.waitConsumeOffQueue.Total(),
+	}).Debug("EventQueue event processing statistics")
+}
+
+// Run consumes events that have been queued for this EventQueue. It
+// is presumed that the eventQueue is a buffered channel with a length of one
+// (i.e., only one event can be processed at a time). All business logic for
+// handling queued events is contained within this function. If the Metadata
+// within an Event does not implement the EventHandler interface, the function
+// will log an error, and the Event will not be processed. If the event queue is
+// closed, then all events which were queued up, but not processed, are
+// cancelled. Any event which is currently being processed will not be
+// cancelled. It is assumed that the caller handles both cases for an Event
+// (cancel, or result of event) gracefully.
+func (q *EventQueue) Run() {
+	go q.eventQueueOnce.Do(func() {
+		for {
+			select {
+			// Receive next event. No other goroutine or process should consume
+			// events off of this channel!
+			case ev := <-q.events:
+				{
+					ev.stats.waitConsumeOffQueue.End(true)
+					eventHandler := ev.Metadata.(EventHandler)
+					ev.stats.durationStat.Start()
+					eventHandler.Handle(ev.eventResults)
+					// Always indicate success for now.s
+					ev.stats.durationStat.End(true)
+					// Ensures that no more results can be sent as the event has
+					// already been processed.
+					ev.printStats()
+					close(ev.eventResults)
+				}
+			// Cancel all events that were not yet consumed.
+			case <-q.close:
+				{
+					// Drain queue of all events.
+					for {
+						select {
+						case drainEvent := <-q.events:
+							drainEvent.stats.waitConsumeOffQueue.End(false)
+							close(drainEvent.cancelled)
+							close(drainEvent.eventResults)
+							drainEvent.printStats()
+						default:
+							// No more events are in events channel, so we can close
```

Comment:
This is not guaranteed, if you close `q.events` we will panic on line 156 since the channel was closed.

Reply:
I thought this would be guaranteed, due to the following:

* `q.close` is closed here because we are in the `case <-q.close:` case within the `select` statement.
* `Enqueue` checks if `q.close` is closed before enqueueing.

Are you envisioning the following scenario?

1. `q.close` is not closed yet
2.`Enqueue` is called, so it `default` case
3. Right after `default` case is reached, `q.close` is closed.
4. The `case <-q.close` is hit in the `select` statement in `q.Run`
5. The queue is drained, and `q.events` is. closed
6. The `Enqueue` call from 2. then tries to write to the channel, and panic occurs.

Reply:
I have been wracking my brain with a way to ensure that this panic scenario doesn't happen while notifying events that they have been cancelled once the queue is closed. I could simply never close the `events` channel, and leave the management of it to the garbage collector, but the problem in that case is that there will be nothing to signal to the event that it will not be processed, as the event will be stuck in the `events` channel forever. Anything that is waiting on its `results` channel will be blocked forever as a result. 

I thought about adding a mutex which blocks out the `case <-q.close` within `q.Run()` from the `Enqueue` operation, but the order in which locks are acquired by waiters in golang is not performed in a FIFO manner per my reading of the documentation. 

I wonder if doing something with some type of `sync.WaitGroup` would allow me to achieve my goal...

Reply:
@aanm I've added use of a `sync.WaitGroup` as well as additional logic with channels in the last commit of the PR. I plan on folding this into the `eventqueue` commit before merging, but wanted to make it separate for easier reviewing. Let me know what you think.